server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java

```diff
@@ -423,7 +425,10 @@ public static void registerRequestHandler(TransportService transportService, Sea
             DFS_ACTION_NAME,
             EsExecutors.DIRECT_EXECUTOR_SERVICE,
             ShardSearchRequest::new,
-            (request, channel, task) -> searchService.executeDfsPhase(request, (SearchShardTask) task, new ChannelActionListener<>(channel))
+            (request, channel, task) -> instrumentedHandler(
+                DFS_ACTION_NAME,
```

Comment:
I just realized that the overall notion that these are measured on the shard, is not entirely accurate. can match is a case where we group shard requests per data node. Hence the timing here would be "how long does it take for can match to be executed on a certain data node". It is still a good metric, but it covers more than one shard. For individual can match on each shard, we need to use a different mechanism. Perhaps we need to adjust the name of the metric then and keep it more general. We could also reuse it for coord level timing, and perhaps the action name needs to include the relevant context?

Also, I would use a more readable action name for our metrics. Some of these action names are hard to distinguish from one another. It took me a bit to trace back all the relevant info, hence I am adding some comments around each action.

This is the can match phase which is executed to pre filter shards that a search request hits. It rewrites the query on the shard and checks whether the result of the rewrite matches no documents, in which case the shard can be filtered out:

- QUERY_CAN_MATCH_NODE_NAME = "shards_can_match";

These two are specific to DFS_QUERY_THEN_FETCH that fetches distributed term frequencies, and executes KNN (the top level search element):
- DFS_ACTION_NAME: dfs_query_then_fetch/shard_dfs_phase
- QUERY_ID_ACTION_NAME = dfs_query_then_fetch/shard_query_phase

This it the plain old query phase if no DFS and KNN or scroll are requested:
- QUERY_ACTION_NAME = query_then_fetch/shard_query_phase

- FREE_CONTEXT_ACTION_NAME: shard_release_context
- 
This one is shared by query_then_fetch and dfs_query_then_fetch
- FETCH_ID_ACTION_NAME: shard_fetch_phase


These three are scroll specific: query then fetch, or a shortcut scenario where query and fetch are done in the shard as part of a single roundtrip:
- QUERY_SCROLL_ACTION_NAME: scroll/shard_query_phase
- FETCH_ID_SCROLL_ACTION_NAME = scroll/shard_fetch_phase
- QUERY_FETCH_SCROLL_ACTION_NAME = scroll/shard_query_and_fetch_phase
- FREE_CONTEXT_SCROLL_ACTION_NAME: scroll/shard_release_context
- CLEAR_SCROLL_CONTEXTS_ACTION_NAME: scroll/shard_release_contexts



I would also add metrics for the coordinator part of these phases:

- coord/can_match
- dfs_query_then_fetch/coord_dfs_phase
- dfs_query_then_fetch/coord_query_phase
- query_then_fetch/coord_query_phase
- coord_fetch_phase
- scroll/coord_query_phase
- scroll/coord_fetch_phase
- scroll/coord_query_and_fetch_phase
- scroll/coord_release_context
- scroll/coord_release_contexts
- coord_release_context


Does the naming make sense? Do the additional coord metrics sound good to you?

Reply:
Excellent comment, thanks for the time you took to write this!

>I just realized that the overall notion that these are measured on the shard, is not entirely accurate. can match is a case where we group shard requests per data node.

To this effect, should I remove the `shard_*` references in your descriptions then? e.g. ` "shards_can_match"` -> ` "can_match"`

Reply:
Ah, I missunderstood.  Reverted.